docs/sql-data-sources-protobuf.md

```diff
@@ -0,0 +1,384 @@
+format, and `from_protobuf()` to decode protobuf binary data into a column. Both functions transform one column to
+another column, and the input/output SQL data type can be a complex type or a primitive type.
+
+Using protobuf message as columns is useful when reading from or writing to a streaming source like Kafka. Each
+Kafka key-value record will be augmented with some metadata, such as the ingestion timestamp into Kafka, the offset in Kafka, etc.
+* If the "value" field that contains your data is in protobuf, you could use `from_protobuf()` to extract your data, enrich it, clean it, and then push it downstream to Kafka again or write it out to a different sink.
+* `to_protobuf()` can be used to turn structs into protobuf message. This method is particularly useful when you would like to re-encode multiple columns into a single one when writing data out to Kafka.
+
+Spark SQL schema is generated based on the protobuf descriptor file or protobuf class passed to `from_protobuf` and `to_protobuf`. The specified protobuf class or protobuf descriptor file must match the data, otherwise, the behavior is undefined: it may fail or return arbitrary results.
+
+<div class="codetabs">
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+import org.apache.spark.sql.protobuf.functions._
+
+// `from_protobuf` and `to_protobuf` provides two schema choices. First, via the protobuf descriptor
+// file, and then via the protobuf message class name.
+// give input .proto protobuf schema
+// syntax  = "proto3"
+// message AppEvent {
+//    string name = 1;
+//    int64 id = 2;
+//    string context = 3;
+// }
+
+val df = spark
+.readStream
+.format("kafka")
+.option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+.option("subscribe", "topic1")
+.load()
+
+// 1. Decode the Protobuf data of schema `AppEvent` into a struct;
+// 2. Filter by column `name`;
+// 3. Encode the column `event` in Protobuf format.
+// The Protobuf protoc command can be used to generate a protobuf descriptor file for give .proto file.
+val output = df
+.select(from_protobuf($"value", "AppEvent", descriptorFilePath) as $"event")
+.where("event.name == \"alice\"")
+.select(to_protobuf($"user", "AppEvent", descriptorFilePath) as $"event")
+
+val query = output
+.writeStream
+.format("kafka")
+.option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+.option("topic", "topic2")
+.start()
+
+// Alternatively, you can decode and encode the SQL columns into protobuf format using protobuf
+// class name. The specified Protobuf class must match the data, otherwise the behavior is undefined:
+// it may fail or return arbitrary result. The jar containing Java class should be shaded.
+// Specifically, `com.google.protobuf.*` should be shaded to `org.sparkproject.spark-protobuf.protobuf.*`.
+var output = df
+.select(from_protobuf($"value", "org.sparkproject.spark-protobuf.protobuf.AppEvent") as $"event")
```

Comment:
I created a small github project to help with shading. https://github.com/rangadi/shaded-protobuf-classes 
We can mention that here. 
In addition, what is shaded are the com.goggle.protobuf classes (like `Message`). Not the user Protobuf classes. We don't need shaded prefix for `AppEvent`. It will be like the following (with the example from `shaded-protobuf-classes` project)
```scala
   .select(from_protobuf($"value", "org.example.protos.AppEvent") as $"event")
```